File: code/go/ch17/src/luago/state/api_arith.go
package state

import "math"
import . "luago/api"
import "luago/number"

type operator struct {
	metamethod  string
	integerFunc func(int64, int64) int64
	floatFunc   func(float64, float64) float64
}

var (
	iadd  = func(a, b int64) int64 { return a + b }
	fadd  = func(a, b float64) float64 { return a + b }
	isub  = func(a, b int64) int64 { return a - b }
	fsub  = func(a, b float64) float64 { return a - b }
	imul  = func(a, b int64) int64 { return a * b }
	fmul  = func(a, b float64) float64 { return a * b }
	imod  = number.IMod
	fmod  = number.FMod
	pow   = math.Pow
	div   = func(a, b float64) float64 { return a / b }
	iidiv = number.IFloorDiv
	fidiv = number.FFloorDiv
	band  = func(a, b int64) int64 { return a & b }
	bor   = func(a, b int64) int64 { return a | b }
	bxor  = func(a, b int64) int64 { return a ^ b }
	shl   = number.ShiftLeft
	shr   = number.ShiftRight
	iunm  = func(a, _ int64) int64 { return -a }
	funm  = func(a, _ float64) float64 { return -a }
	bnot  = func(a, _ int64) int64 { return ^a }
)

var operators = []operator{
	operator{"__add", iadd, fadd},
	operator{"__sub", isub, fsub},
	operator{"__mul", imul, fmul},
	operator{"__mod", imod, fmod},
	operator{"__pow", nil, pow},
	operator{"__div", nil, div},
	operator{"__idiv", iidiv, fidiv},
	operator{"__band", band, nil},
	operator{"__bor", bor, nil},
	operator{"__bxor", bxor, nil},
	operator{"__shl", shl, nil},
	operator{"__shr", shr, nil},
	operator{"__unm", iunm, funm},
	operator{"__bnot", bnot, nil},
}

// [-(2|1), +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_arith
func (self *luaState) Arith(op ArithOp) {
	var a, b luaValue // operands
	b = self.stack.pop()
	if op != LUA_OPUNM && op != LUA_OPBNOT {
		a = self.stack.pop()
	} else {
		a = b
	}

	operator := operators[op]
	if result := _arith(a, b, operator); result != nil {
		self.stack.push(result)
		return
	}

	mm := operator.metamethod
	if result, ok := callMetamethod(a, b, mm, self); ok {
		self.stack.push(result)
		return
	}

	panic("arithmetic error!")
}

func _arith(a, b luaValue, op operator) luaValue {
	if op.floatFunc == nil { // bitwise
		if x, ok := convertToInteger(a); ok {
			if y, ok := convertToInteger(b); ok {
				return op.integerFunc(x, y)
			}
		}
	} else { // arith
		if op.integerFunc != nil { // add,sub,mul,mod,idiv,unm
			if x, ok := a.(int64); ok {
				if y, ok := b.(int64); ok {
					return op.integerFunc(x, y)
				}
			}
		}
		if x, ok := convertToFloat(a); ok {
			if y, ok := convertToFloat(b); ok {
				return op.floatFunc(x, y)
			}
		}
	}
	return nil
}

File: code/csharp/ch06/luavm/state/ApiArith.cs
using System;
using luavm.api;
using ArithOp = System.Int32;
using CompareOp = System.Int32;

namespace luavm.state
{
    internal delegate long IntegerFunc(long a, long b);

    internal delegate double FloatFunc(double a, double b);

    struct Operator
    {
        public IntegerFunc integerFunc;
        public FloatFunc floatFunc;
    }

    public partial struct LuaState
    {
        private static Operator[] operators = new Operator[]
        {
            new Operator
            {
                integerFunc = iadd,
                floatFunc = fadd
            },
            new Operator
            {
                integerFunc = isub,
                floatFunc = fsub
            },
            new Operator
            {
                integerFunc = imul,
                floatFunc = fmul
            },
            new Operator
            {
                integerFunc = imod,
                floatFunc = fmod
            },
            new Operator
            {
                integerFunc = null,
                floatFunc = pow
            },
            new Operator
            {
                integerFunc = null,
                floatFunc = div
            },
            new Operator
            {
                integerFunc = iidiv,
                floatFunc = fidiv
            },
            new Operator
            {
                integerFunc = band,
                floatFunc = null
            },
            new Operator
            {
                integerFunc = bor,
                floatFunc = null
            },
            new Operator
            {
                integerFunc = bxor,
                floatFunc = null
            },
            new Operator
            {
                integerFunc = shl,
                floatFunc = null
            },
            new Operator
            {
                integerFunc = shr,
                floatFunc = null
            },
            new Operator
            {
                integerFunc = inum,
                floatFunc = fnum
            },
            new Operator
            {
                integerFunc = bnot,
                floatFunc = null
            }
        };

        private static long iadd(long a, long b)
        {
            return a + b;
        }

        private static double fadd(double a, double b)
        {
            return a + b;
        }

        private static long isub(long a, long b)
        {
            return a - b;
        }

        private static double fsub(double a, double b)
        {
            return a - b;
        }

        private static long imul(long a, long b)
        {
            return a * b;
        }

        private static double fmul(double a, double b)
        {
            return a * b;
        }

        private static long imod(long a, long b)
        {
            return number.Math.IMod(a, b);
        }

        private static double fmod(double a, double b)
        {
            return number.Math.FMod(a, b);
        }

        private static double pow(double a, double b)
        {
            return System.Math.Pow(a, b);
        }

        private static double div(double a, double b)
        {
            return a / b;
        }

        private static long iidiv(long a, long b)
        {
            return number.Math.IFloorDiv(a, b);
        }

        private static double fidiv(double a, double b)
        {
            return number.Math.FFloorDiv(a, b);
        }

        private static long band(long a, long b)
        {
            return a & b;
        }

        private static long bor(long a, long b)
        {
            return a | b;
        }

        private static long bxor(long a, long b)
        {
            return a ^ b;
        }

        private static long shl(long a, long b)
        {
            return number.Math.ShiftLeft(a, b);
        }

        private static long shr(long a, long b)
        {
            return number.Math.ShiftRight(a, b);
        }

        private static long inum(long a, long _)
        {
            return -a;
        }

        private static double fnum(double a, double _)
        {
            return -a;
        }

        private static long bnot(long a, long _)
        {
            //书上是^a
            return ~a;
        }


        public void Arith(ArithOp op)
        {
            LuaValue a, b;
            b = new LuaValue(stack.pop());
            if (op != Consts.LUA_OPUNM && op != Consts.LUA_OPBNOT)
            {
                a = new LuaValue(stack.pop());
            }
            else
            {
                a = b;
            }

            var opr = operators[op];
            var result = _arith(a, b, opr);
            if (result != null)
            {
                stack.push(result);
            }
            else
            {
                throw new Exception("arithmetic error!");
            }
        }

        object _arith(LuaValue a, LuaValue b, Operator op)
        {
            if (op.floatFunc == null)
            {
                Tuple<long, bool> v = LuaValue.convertToInteger(a.value);
                if (v.Item2)
                {
                    Tuple<long, bool> v2 = LuaValue.convertToInteger(b.value);
                    if (v2.Item2)
                    {
                        return op.integerFunc(v.Item1, v2.Item1);
                    }
                }
            }
            else
            {
                if (op.integerFunc != null)
                {
                    if (a.value.GetType().Name.Equals("Int64") && b.value.GetType().Name.Equals("Int64"))
                    {
                        var x = long.Parse(a.value.ToString());
                        var y = long.Parse(b.value.ToString());
                        return op.integerFunc(x, y);
                    }
                }

                var v = LuaValue.convertToFloat(a.value);
                if (v.Item2)
                {
                    var v2 = LuaValue.convertToFloat(b.value);
                    if (v2.Item2)
                    {
                        return op.floatFunc(v.Item1, v2.Item1);
                    }
                }
            }

            return null;
        }
    }
}